src/gc/collector.cpp

```diff
@@ -115,8 +140,12 @@ class TraceStack {
             pop_chunk();
             assert(cur == end);
             return *--cur; // no need for any bounds checks here since we're guaranteed we're CHUNK_SIZE from the start
+        } else {
+            // We emptied the stack, but we should prepare a new chunk in case another item
```

Comment:
Yes, because in the mark phase I added an extra step where some classes get added to the TraceStack after one traversal, and then do another traversal, so I'm reusing the same TraceStack. I could also create another TraceStack, but this makes the TraceStack implementation more "correct" and it was pretty hard to figure out that the problem was in the TraceStack because it would manifest as memory corruption elsewhere in the program.